File: distribution/python/dqops/client/models/rule_thresholds_model.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.rule_parameters_model import RuleParametersModel


T = TypeVar("T", bound="RuleThresholdsModel")


@_attrs_define
class RuleThresholdsModel:
    """Model that returns the form definition and the form data to edit a single rule with all three threshold levels (low,
    medium, high).

        Attributes:
            error (Union[Unset, RuleParametersModel]): Model that returns the form definition and the form data to edit
                parameters (thresholds) for a rule at a single severity level (low, medium, high).
            warning (Union[Unset, RuleParametersModel]): Model that returns the form definition and the form data to edit
                parameters (thresholds) for a rule at a single severity level (low, medium, high).
            fatal (Union[Unset, RuleParametersModel]): Model that returns the form definition and the form data to edit
                parameters (thresholds) for a rule at a single severity level (low, medium, high).
    """

    error: Union[Unset, "RuleParametersModel"] = UNSET
    warning: Union[Unset, "RuleParametersModel"] = UNSET
    fatal: Union[Unset, "RuleParametersModel"] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        error: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.error, Unset):
            error = self.error.to_dict()

        warning: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.warning, Unset):
            warning = self.warning.to_dict()

        fatal: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.fatal, Unset):
            fatal = self.fatal.to_dict()

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if error is not UNSET:
            field_dict["error"] = error
        if warning is not UNSET:
            field_dict["warning"] = warning
        if fatal is not UNSET:
            field_dict["fatal"] = fatal

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.rule_parameters_model import RuleParametersModel

        d = src_dict.copy()
        _error = d.pop("error", UNSET)
        error: Union[Unset, RuleParametersModel]
        if isinstance(_error, Unset):
            error = UNSET
        else:
            error = RuleParametersModel.from_dict(_error)

        _warning = d.pop("warning", UNSET)
        warning: Union[Unset, RuleParametersModel]
        if isinstance(_warning, Unset):
            warning = UNSET
        else:
            warning = RuleParametersModel.from_dict(_warning)

        _fatal = d.pop("fatal", UNSET)
        fatal: Union[Unset, RuleParametersModel]
        if isinstance(_fatal, Unset):
            fatal = UNSET
        else:
            fatal = RuleParametersModel.from_dict(_fatal)

        rule_thresholds_model = cls(
            error=error,
            warning=warning,
            fatal=fatal,
        )

        rule_thresholds_model.additional_properties = d
        return rule_thresholds_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties

File: dqops/src/main/java/com/dqops/services/check/mapping/models/RuleThresholdsModel.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */
package com.dqops.services.check.mapping.models;

import com.dqops.utils.exceptions.DqoRuntimeException;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import io.swagger.annotations.ApiModel;
import lombok.Data;

/**
 * Model that returns the form definition and the form data to edit a single rule with all three threshold levels (low, medium, high).
 */
@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@ApiModel(value = "RuleThresholdsModel", description = "Model that returns the form definition and the form data to edit a single rule with all three threshold levels (low, medium, high).")
public class RuleThresholdsModel implements Cloneable {
    /**
     * Rule parameters for the error severity rule.
     */
    @JsonPropertyDescription("Rule parameters for the error severity rule.")
    private RuleParametersModel error;

    /**
     * Rule parameters for the warning severity rule.
     */
    @JsonPropertyDescription("Rule parameters for the warning severity rule.")
    private RuleParametersModel warning;

    /**
     * Rule parameters for the fatal severity rule.
     */
    @JsonPropertyDescription("Rule parameters for the fatal severity rule.")
    private RuleParametersModel fatal;

    /**
     * Finds the first not-null rule that is used for generating the documentation.
     * @return The first not null rule.
     */
    public RuleParametersModel findFirstNotNullRule() {
        if (this.error != null) {
            return this.error;
        }

        if (this.warning != null) {
            return this.warning;
        }

        return this.fatal;
    }

    /**
     * Creates a selective deep/shallow clone of the object. Definition objects are not cloned, but all other editable objects are.
     * @return Cloned instance.
     */
    public RuleThresholdsModel cloneForUpdate() {
        try {
            RuleThresholdsModel cloned = (RuleThresholdsModel) super.clone();
            if (cloned.error != null) {
                cloned.error = cloned.error.cloneForUpdate();
            }
            if (cloned.warning != null) {
                cloned.warning = cloned.warning.cloneForUpdate();
            }
            if (cloned.fatal != null) {
                cloned.fatal = cloned.fatal.cloneForUpdate();
            }

            return cloned;
        }
        catch (CloneNotSupportedException ex) {
            throw new DqoRuntimeException("Clone not supported: " + ex.toString(), ex);
        }
    }

    /**
     * Applies sample values for fields that have a sample value. Overrides the current values.
     * The model filled with sample values is used for generating the documentation model.
     */
    public void applySampleValues() {
        if (this.error != null) {
            this.error.applySampleValues();
        }
        if (this.warning != null) {
            this.warning.applySampleValues();
        }
        if (this.fatal != null) {
            this.fatal.applySampleValues();
        }
    }

    /**
     * Checks if any rule at any severity level is configured (enabled).
     * @return One of the rules is enabled.
     */
    public boolean hasAnyRulesConfigured() {
        return this.warning.isConfigured() || this.error.isConfigured() || this.fatal.isConfigured();
    }
}
